[SDK 50] Regression: expo-sqlite no longer supports FTS modules

### Minimal reproducible example

https://github.com/ndekross/expo-sqlite-no-fts-module-error

### Summary

We've been using the SQlite's [FTS4 module](https://www.sqlite.org/fts3.html) in our app, which is currently running on Expo 49 and `expo-sqlite@11.3.3`. This module is crucial for providing full-text search, a core feature for our customers. However, after upgrading to Expo SDK 50 and updating `expo-sqlite` to the latest version, the FTS module has stopped functioning. When I attempt to create an FTS table, SQLite gives this error:

```
 [Error: Error code 1: no such module: fts4]
```

I also tried FTS3 and FTS5, but neither was successful. I've created a [minimal reproducible example](https://github.com/ndekross/expo-sqlite-no-fts-module-error) based on Expo's `with-sqlite` template. You can start the app and observe the error firsthand.

What we expect:
- Error-free creation of an FTS4 virtual table
- Support for FTS5, which was released nearly a decade ago

Thank you!

### Environment

```
expo-env-info 1.2.0 environment info:
    System:
      OS: Windows 11 10.0.22621
    Binaries:
      Node: 18.12.1 - C:\Program Files\nodejs\node.EXE
      Yarn: 3.6.3 - ~\AppData\Roaming\npm\yarn.CMD
      npm: 9.6.6 - ~\AppData\Roaming\npm\npm.CMD
    IDEs:
      Android Studio: AI-231.9392.1.2311.11255304
    npmPackages:
      expo: ^50.0.4 => 50.0.5
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.73.2 => 0.73.2
      react-native-web: ~0.19.8 => 0.19.8
    Expo Workflow: bare
```

expo-sqlite/next
I am having a nightmare with bulk insertions alone or in a loop to import 31200 records. Take them from the server and to insert into the table it is a nightmare.

 [Error: Call to function 'NativeDatabase.execAsync' has been rejected.
→ Caused by: java.lang.NullPointerException: java.lang.NullPointerException]

after somme correct insertions inside a table table it suddenly give this error, some times it worked completele when I deleted the schemas to make sure that everything works and try to import data again and booom 
error 

[Error: Call to function 'NativeDatabase.execAsync' has been rejected.
→ Caused by: java.lang.NullPointerException: java.lang.NullPointerException]

some times it does not even insert 1 record. this crazy.

I put my component as PureComponent to make sure it is not beeing disturbed by re-rendering

for this bloc 
let db = await SQLite.openDatabaseAsync('mybiblebooks');   
await db.execAsync(`
	PRAGMA journal_mode = WAL;
	${sqlRecordstxt}
	`);
	
it worked some times and sometimes gives the same error 	
[Error: Call to function 'NativeDatabase.execAsync' has been rejected.
→ Caused by: java.lang.NullPointerException: java.lang.NullPointerException]

Note: the data during insertion sometimes are Latin chars 				  
				  
//************************************************************************************************************/				  

			try {  
					let db = await SQLite.openDatabaseAsync('mybiblebooks');   
				  //   await db.execAsync(`
				  //   PRAGMA journal_mode = WAL;
				  //   ${sqlRecordstxt}
				  //   `);
				  let qtyGo=0;
				  for await (let versbloc of response.verslist){  
					  qtyGo++;
					  try{
						  //const verslist= await this.bulkOperationOff(db, qtyGo, versbloc); 
						  //const db = await SQLite.openDatabaseAsync('mybiblebooks');  
							// PRAGMA journal_mode = WAL; 
							///setTimeout(()=>{}, 26000);
							//this2.setTimeout(()=>{}, 26000);

							await db.execAsync(`
							${versbloc}
							`);
							//const result=await db.runAsync('INSERT INTO vers VALUES (?, ?, ?, ?, ?, ?, ?)', versbloc); //succed during 108 insertions then error 
							// [Error: Call to function 'NativeDatabase.execAsync' has been rejected.
							//→ Caused by: java.lang.NullPointerException: java.lang.NullPointerException]

							console.log("__|__success INSERT operation______ in bulkOperation #", qtyGo);
							//await db.closeAsync();
						  	console.log("*******************NOTES-POSTACTION-siiiiize**********00000**********verslist2 ", qtyGo, versbloc.length);
							///  this2.setTimeout(()=>{}, 26000);
					  }catch (error00) {
						  console.log("error bulkOperationOffx0000000000000000000---00000000 -tooooooopnow setp=", qtyGo, " size=", versbloc.length);
						  //console.log("error Arr ", versbloc);
						  console.log(error00);
						  //return "error";
						  // if(error00.toString().indexOf("NullPointerException")>-1){
						  // 	db = await SQLite.openDatabaseAsync('mybiblebooks'); 
						  // 	const verslist2= await this.bulkOperationOff(db, qtyGo, versbloc); 
						  // 	console.log("*******************CATCH***ERROR-NOTES-POSTACTION-siiiiize**********00000**********verslist2 ", qtyGo, verslist2);  
						  // }
					  }
						
					  //await new Promise(resolve => setTimeout(resolve, 1000)); //26000 
				  }
	  
				console.log("success operation in bulkOperationOffx ");
				//await db.closeAsync();
	  
				//return "yes";
			  }
			  catch (error) {
				  console.log("error bulkOperationOffx tooooooopnow ");
				  console.log(error);
				  //return "error";
			  }



The errors after some succeeded insertions 

[mass-insert-31200-records.txt](https://github.com/expo/expo/files/14201743/mass-insert-31200-records.txt)

> expo-sqlite/next I am having a nightmare with bulk insertions alone or in a loop to import 31200 records. Take them from the server and to insert into the table it is a nightmare.
> 
> [Error: Call to function 'NativeDatabase.execAsync' has been rejected. → Caused by: java.lang.NullPointerException: java.lang.NullPointerException]
> 
> after somme correct insertions inside a table table it suddenly give this error, some times it worked completele when I deleted the schemas to make sure that everything works and try to import data again and booom error
> 
> [Error: Call to function 'NativeDatabase.execAsync' has been rejected. → Caused by: java.lang.NullPointerException: java.lang.NullPointerException]
> 
> some times it does not even insert 1 record. this crazy.
> 
> I put my component as PureComponent to make sure it is not beeing disturbed by re-rendering
> 
> for this bloc let db = await SQLite.openDatabaseAsync('mybiblebooks'); await db.execAsync(`PRAGMA journal_mode = WAL; ${sqlRecordstxt}`);
> 
> it worked some times and sometimes gives the same error [Error: Call to function 'NativeDatabase.execAsync' has been rejected. → Caused by: java.lang.NullPointerException: java.lang.NullPointerException]
> 
> Note: the data during insertion sometimes are Latin chars
> 
> //************************************************************************************************************/
> 
> ```
> 		try {  
> 				let db = await SQLite.openDatabaseAsync('mybiblebooks');   
> 			  //   await db.execAsync(`
> 			  //   PRAGMA journal_mode = WAL;
> 			  //   ${sqlRecordstxt}
> 			  //   `);
> 			  let qtyGo=0;
> 			  for await (let versbloc of response.verslist){  
> 				  qtyGo++;
> 				  try{
> 					  //const verslist= await this.bulkOperationOff(db, qtyGo, versbloc); 
> 					  //const db = await SQLite.openDatabaseAsync('mybiblebooks');  
> 						// PRAGMA journal_mode = WAL; 
> 						///setTimeout(()=>{}, 26000);
> 						//this2.setTimeout(()=>{}, 26000);
> 
> 						await db.execAsync(`
> 						${versbloc}
> 						`);
> 						//const result=await db.runAsync('INSERT INTO vers VALUES (?, ?, ?, ?, ?, ?, ?)', versbloc); //succed during 108 insertions then error 
> 						// [Error: Call to function 'NativeDatabase.execAsync' has been rejected.
> 						//→ Caused by: java.lang.NullPointerException: java.lang.NullPointerException]
> 
> 						console.log("__|__success INSERT operation______ in bulkOperation #", qtyGo);
> 						//await db.closeAsync();
> 					  	console.log("*******************NOTES-POSTACTION-siiiiize**********00000**********verslist2 ", qtyGo, versbloc.length);
> 						///  this2.setTimeout(()=>{}, 26000);
> 				  }catch (error00) {
> 					  console.log("error bulkOperationOffx0000000000000000000---00000000 -tooooooopnow setp=", qtyGo, " size=", versbloc.length);
> 					  //console.log("error Arr ", versbloc);
> 					  console.log(error00);
> 					  //return "error";
> 					  // if(error00.toString().indexOf("NullPointerException")>-1){
> 					  // 	db = await SQLite.openDatabaseAsync('mybiblebooks'); 
> 					  // 	const verslist2= await this.bulkOperationOff(db, qtyGo, versbloc); 
> 					  // 	console.log("*******************CATCH***ERROR-NOTES-POSTACTION-siiiiize**********00000**********verslist2 ", qtyGo, verslist2);  
> 					  // }
> 				  }
> 					
> 				  //await new Promise(resolve => setTimeout(resolve, 1000)); //26000 
> 			  }
>   
> 			console.log("success operation in bulkOperationOffx ");
> 			//await db.closeAsync();
>   
> 			//return "yes";
> 		  }
> 		  catch (error) {
> 			  console.log("error bulkOperationOffx tooooooopnow ");
> 			  console.log(error);
> 			  //return "error";
> 		  }
> ```
> 
> The errors after some succeeded insertions
> 
> [mass-insert-31200-records.txt](https://github.com/expo/expo/files/14201743/mass-insert-31200-records.txt)

I have the same random issue that blocks everything. Is it a known issue?

--- a/packages/expo-sqlite/CHANGELOG.md
+++ b/packages/expo-sqlite/CHANGELOG.md
@@ -8,6 +8,8 @@
 
 ### 🐛 Bug fixes
 
+- Enabled [FTS](https://www.sqlite.org/fts3.html) and [FTS5](https://www.sqlite.org/fts5.html) for SQLite. ([#27738](https://github.com/expo/expo/pull/27738) by [@kudo](https://github.com/kudo))
+
 ### 💡 Others
 
 - drop unused web `name` property. ([#27437](https://github.com/expo/expo/pull/27437) by [@EvanBacon](https://github.com/EvanBacon))

--- a/packages/expo-sqlite/android/CMakeLists.txt
+++ b/packages/expo-sqlite/android/CMakeLists.txt
@@ -10,11 +10,10 @@ set(BUILD_DIR ${CMAKE_SOURCE_DIR}/build)
 set(SRC_DIR "${CMAKE_SOURCE_DIR}/src/main/cpp")
 file(GLOB SOURCES "${SRC_DIR}/*.cpp")
 
-if (NOT "${SQLITE_CUSTOM_BUILDFLAGS}" STREQUAL "")
-  add_compile_options(
-    "${SQLITE_CUSTOM_BUILDFLAGS}"
-  )
-endif()
+separate_arguments(SQLITE_BUILDFLAGS)
+add_compile_options(
+  ${SQLITE_BUILDFLAGS}
+)
 
 add_library(
   ${PACKAGE_NAME}

--- a/packages/expo-sqlite/android/build.gradle
+++ b/packages/expo-sqlite/android/build.gradle
@@ -32,6 +32,20 @@ def customDownloadsDir = System.getenv("REACT_NATIVE_DOWNLOADS_DIR")
 def downloadsDir = customDownloadsDir ? new File(customDownloadsDir) : new File("$buildDir/downloads")
 def SQLITE3_SRC_DIR = new File("$buildDir/sqlite3_src")
 
+def getSQLiteBuildFlags() {
+  def buildFlags = ''
+  if (findProperty('expo.sqlite.enableFTS') !== 'false') {
+    buildFlags <<= '-DSQLITE_ENABLE_FTS4=1 -DSQLITE_ENABLE_FTS3_PARENTHESIS=1 -DSQLITE_ENABLE_FTS5=1'
+  } else {
+  }
+  def customBuildFlags = findProperty('expo.sqlite.customBuildFlags') ?: ''
+  if (customBuildFlags != '') {
+    buildFlags <<= " ${customBuildFlags}"
+  }
+  logger.info("SQLite build flags: ${buildFlags}")
+  return buildFlags
+}
+
 def reactNativeArchitectures() {
   def value = project.getProperties().get("reactNativeArchitectures")
   return value ? value.split(",") : ["armeabi-v7a", "x86", "x86_64", "arm64-v8a"]
@@ -123,7 +137,7 @@ android {
         abiFilters (*reactNativeArchitectures())
         arguments "-DANDROID_STL=c++_shared",
           "-DSQLITE3_SRC_DIR=${toPlatformIndependentPath(SQLITE3_SRC_DIR)}",
-          "-DSQLITE_CUSTOM_BUILDFLAGS=${findProperty('expo.sqlite.customBuildFlags') ?: ''}"
+          "-DSQLITE_BUILDFLAGS=${getSQLiteBuildFlags()}"
       }
     }
   }

--- a/packages/expo-sqlite/ios/ExpoSQLite.podspec
+++ b/packages/expo-sqlite/ios/ExpoSQLite.podspec
@@ -1,6 +1,9 @@
 require 'json'
 
 package = JSON.parse(File.read(File.join(__dir__, '..', 'package.json')))
+podfile_properties = JSON.parse(File.read("#{Pod::Config.instance.installation_root}/Podfile.properties.json")) rescue {}
+
+SQLITE_VERSION = '3.42.0'
 
 Pod::Spec.new do |s|
   s.name           = 'ExpoSQLite'
@@ -14,8 +17,12 @@ Pod::Spec.new do |s|
   s.source         = { git: 'https://github.com/expo/expo.git' }
   s.static_framework = true
   s.dependency 'ExpoModulesCore'
-  # The builtin sqlite does not support extensions so we update it
-  s.dependency 'sqlite3', '~> 3.42.0'
+
+  s.dependency 'sqlite3', "~> #{SQLITE_VERSION}"
+  unless podfile_properties['expo.sqlite.enableFTS'] === 'false'
+    s.dependency 'sqlite3/fts', "~> #{SQLITE_VERSION}"
+    s.dependency 'sqlite3/fts5', "~> #{SQLITE_VERSION}"
+  end
 
   # Swift/Objective-C compatibility
   s.pod_target_xcconfig = {

--- a/packages/expo-sqlite/ios/ExpoSQLite.podspec
+++ b/packages/expo-sqlite/ios/ExpoSQLite.podspec
@@ -3,7 +3,7 @@ require 'json'
 package = JSON.parse(File.read(File.join(__dir__, '..', 'package.json')))
 podfile_properties = JSON.parse(File.read("#{Pod::Config.instance.installation_root}/Podfile.properties.json")) rescue {}
 
-SQLITE_VERSION = '3.42.0'
+sqliteVersion = '3.42.0'
 
 Pod::Spec.new do |s|
   s.name           = 'ExpoSQLite'
@@ -18,10 +18,10 @@ Pod::Spec.new do |s|
   s.static_framework = true
   s.dependency 'ExpoModulesCore'
 
-  s.dependency 'sqlite3', "~> #{SQLITE_VERSION}"
+  s.dependency 'sqlite3', "~> #{sqliteVersion}"
   unless podfile_properties['expo.sqlite.enableFTS'] === 'false'
-    s.dependency 'sqlite3/fts', "~> #{SQLITE_VERSION}"
-    s.dependency 'sqlite3/fts5', "~> #{SQLITE_VERSION}"
+    s.dependency 'sqlite3/fts', "~> #{sqliteVersion}"
+    s.dependency 'sqlite3/fts5', "~> #{sqliteVersion}"
   end
 
   # Swift/Objective-C compatibility

